src/main/java/com/project/bookapp/service/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,134 @@
+        shoppingCart.addCartItem(savedCartItem);
+
+        shoppingCartRepository.save(shoppingCart);
+
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto findById(Long id) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(id)
+                .orElseThrow(() ->
+                        new EntityNotFoundException("Shopping cart with id " + id + " not found"));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    private ShoppingCart createNewShoppingCart(Long userId) {
+        User user = userRepository.findById(userId)
+                .orElseThrow(()
+                        -> new EntityNotFoundException("User with id " + userId + " not found"));
+        ShoppingCart newCart = new ShoppingCart();
+        newCart.setUser(user);
+        return shoppingCartRepository.save(newCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto update(
+            Long cartItemId,
+            CartItemRequestDto cartItemRequestDto,
+            Long id) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
+                .orElseThrow(()
+                        -> new EntityNotFoundException("Cart item not found with ID: "
+                        + cartItemId));
+
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(id)
+                .orElseThrow(()
+                        -> new EntityNotFoundException("Shopping cart not found for user ID: "
+                        + id));
+
+        if (!cartItem.getShoppingCart().getId().equals(shoppingCart.getId())) {
+            throw new IllegalArgumentException(
+                    "Cart item does not belong to the authenticated user");
+        }
+
+        if (cartItemRequestDto.getBookId() != null
+                && !cartItem.getBook().getId().equals(cartItemRequestDto.getBookId())) {
+            Book book = bookRepository.findById(cartItemRequestDto.getBookId())
+                    .orElseThrow(()
+                            -> new EntityNotFoundException("Book not found with ID: "
+                            + cartItemRequestDto.getBookId()));
+            cartItem.setBook(book);
+        }
+
+        cartItem.setQuantity(cartItemRequestDto.getQuantity());
+
+        cartItemRepository.save(cartItem);
+
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public void deleteById(Long cartItemId, Long userId) {
+
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
+                .orElseThrow(() -> new EntityNotFoundException("Cart item not found with ID: "
+                        + cartItemId));
+
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Shopping cart not found for user ID: " + userId));
+
+        if (!cartItem.getShoppingCart().getId().equals(shoppingCart.getId())) {
```

Comment:
Let’s simplify this method. Find cart item by ID and shopping cart ID on the DB query level